### PR TITLE
Update REST API status code returned when adding files to the wrong type of run

### LINKED ISSUE #178: REST API should return HTTP 400 when adding files to wrong type of run
When adding a single end file to a paired end run (or vice versa) IRIDA returns an HTTP 500 error (server error).  This would make more sense to return a HTTP 400 error (Bad Request).  This is due to throwing an `IllegalArgumentException` in `ca.corefacility.bioinformatics.irida.service.impl.SequencingObjectServiceImpl#create`.  This could be caught in the REST controller and return the appropriate response.

Reported by @jeffrey.thiessen

Imported from GitLab issue #672.  Originally posted on 2018/08/22 02:24PM

### PR DESCRIPTION
**Pull request by deepsidhu85**
## Description of changes
What did you change in this pull request?  Provide a description of files changed, user interactions changed, etc.  Include how to test your changes.

Added try/catch blocks when uploading sequence files to return a bad request status code (400) if an IllegalArgumentException is thrown, or return an Internal Server Error status code (500) for any other exceptions thrown.

To test this:

1) Get the latest version of the IRIDA Uploader
2) Upload a paired end sequence using the uploader
3) Go into the folder which you uploaded the paired end sequence from, open up the irida_uploader_status.info file, and get the run_id
4) Modify the api_handler.py file under the core/ directory as follows:
- Replace this line: `run_id = api_instance.create_seq_run(sequencing_run.metadata)` with `run_id=run_id_retrieved_from_file_above`
5) Now upload a single-end sequence

You should see something similar to:

2019-04-10 15:21:28 ERROR    Error while uploading [08-5578]: [Bad Request]
2019-04-10 15:21:28 ERROR    Failed to upload SequencingRun, Could not connect to IRIDA
2019-04-10 15:21:28 ERROR    Lost connection to Irida
2019-04-10 15:21:28 ERROR    Errors: ('Error 400: Bad Request\n',)

Note: An issue has been made on the Irida Uploader Github page to provide a clear message upon a bad request rather than `Could not connect to IRIDA` and `Lost connection to IRIDA`

## Related issue
Link to the GitHub issue this pull request addresses using the `#issuenum` format.  If it completes an issue, use `Fixes #issuenum` to automatically close the issue.

Fixes #178 

## Checklist
Things for the developer to confirm they've done before the PR should be accepted:

* [x] CHANGELOG.md (and UPGRADING.md if necessary) updated with information for new change.
~~* [ ] Tests added (or description of how to test) for any new features.~~
~~* [ ] User documentation updated for UI or technical changes.~~

### CODE REVIEW

File: src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/samples/RESTSampleSequenceFilesController.java [hunk 1]
```diff
@@ -530,30 +541,38 @@ public ModelMap addNewSequenceFilePairToSample(@PathVariable Long sampleId,
...
+
+			// add location header
+			response.addHeader(HttpHeaders.LOCATION, sequencingObject.getLink("self")
+					.getHref());
+
+			// set the response status.
+			response.setStatus(HttpStatus.CREATED.value());
+			modelMap.addAttribute(RESTGenericController.RESOURCE_NAME, sequencingObject);
+		} catch (IllegalArgumentException e) {
+			response.setStatus(HttpStatus.BAD_REQUEST.value());
```
Comment (tom114) [inline src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/samples/RESTSampleSequenceFilesController.java]: Throw a logging statement here with the exception please so it doesn't just get swallowed.  `debug` level is probably fine.

File: src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/samples/RESTSampleSequenceFilesController.java [hunk 2]
```diff
@@ -422,45 +422,56 @@ public ModelMap addNewSequenceFileToSample(@PathVariable Long sampleId, @Request
...
+			modelMap.addAttribute(RESTGenericController.RESOURCE_NAME, sequenceFile);
+			// add a location header.
+			response.addHeader(HttpHeaders.LOCATION, selfRel.getHref());
+			// set the response status.
+			response.setStatus(HttpStatus.CREATED.value());
+
+		} catch (IllegalArgumentException e) {
+			response.setStatus(HttpStatus.BAD_REQUEST.value());
+		} catch (Exception e) {
+			response.setStatus(HttpStatus.INTERNAL_SERVER_ERROR.value());
```
Comment (tom114) [inline src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/samples/RESTSampleSequenceFilesController.java]: Same comments here about catching `Exception` and logging.

File: src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/samples/RESTSampleSequenceFilesController.java [hunk 3]
```diff
@@ -530,30 +541,38 @@ public ModelMap addNewSequenceFilePairToSample(@PathVariable Long sampleId,
...
+			response.addHeader(HttpHeaders.LOCATION, sequencingObject.getLink("self")
+					.getHref());
+
+			// set the response status.
+			response.setStatus(HttpStatus.CREATED.value());
+			modelMap.addAttribute(RESTGenericController.RESOURCE_NAME, sequencingObject);
+		} catch (IllegalArgumentException e) {
+			response.setStatus(HttpStatus.BAD_REQUEST.value());
+		} catch (Exception e) {
+			response.setStatus(HttpStatus.INTERNAL_SERVER_ERROR.value());
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/samples/RESTSampleSequenceFilesController.java]: Fixed in 1d9c20b27fa35533f4334f521558c3e28baef6ae

File: src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/samples/RESTSampleSequenceFilesController.java (hunk 1 above)
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/samples/RESTSampleSequenceFilesController.java]: Fixed in 1d9c20b27fa35533f4334f521558c3e28baef6ae

File: src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/samples/RESTSampleSequenceFilesController.java (hunk 2 above)
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/web/controller/api/samples/RESTSampleSequenceFilesController.java]: Fixed in 1d9c20b27fa35533f4334f521558c3e28baef6ae

**Comment by tom114** [comment]
Got a CHANGELOG conflict here now.

**Comment by deepsidhu85** [comment]
> Got a CHANGELOG conflict here now.

Fixed in 9ba95fa4631493676bd01c9529f78e829af7aa04

**Pull request merged**